---
title: Azure Active Directory Authentication Library (ADAL) クライアントのエラー処理のベスト プラクティス
description: ADAL クライアント アプリケーションのエラー処理に関するガイダンスとベスト プラクティスを提供します。
services: active-directory
documentationcenter: ''
author: danieldobalian
manager: mtillman
ms.author: celested
ms.service: active-directory
ms.component: develop
ms.devlang: na
ms.topic: article
ms.tgt_pltfrm: na
ms.workload: identity
ms.date: 02/27/2017
ms.custom: ''
ms.openlocfilehash: b28e1931b9f615ae0eebe40b101f1959e9fcb40a
ms.sourcegitcommit: 1f0587f29dc1e5aef1502f4f15d5a2079d7683e9
ms.translationtype: HT
ms.contentlocale: ja-JP
ms.lasthandoff: 08/07/2018
ms.locfileid: "39593045"
---
# <a name="error-handling-best-practices-for-azure-active-directory-authentication-library-adal-clients"></a>Azure Active Directory Authentication Library (ADAL) クライアントのエラー処理のベスト プラクティス

この記事では、ADAL を使ってユーザーを認証するときに開発者側で発生する可能性のあるエラーの種類に関するガイダンスを提供します。 ADAL を使う場合は、開発者がステップ実行してエラーを処理することが必要な可能性のあるケースがいくつかあります。 適切なエラー処理により、エンドユーザー エクスペリエンスが向上し、エンド ユーザーがサインインする必要がある回数が制限されます。

この記事では、ADAL でサポートされている各プラットフォームの特定のケースと、アプリケーションで各ケースを正しく処理する方法について説明します。 エラーのガイダンスは、ADAL API が提供するトークンの取得パターンに基づいて、2 つの大きなカテゴリに分かれています。

- **AcquireTokenSilent**: クライアントはトークンを自動で (UI なし) 取得しようとし、ADAL が失敗した場合は失敗する可能性があります。 
- **AcquireToken**: クライアントは自動での取得を試みることができますが、サインインが必要な対話型の要求を実行することもできます。

> [!TIP]
> ADAL と Azure AD を使う場合は、すべてのエラーと例外をログに記録することをお勧めします。 ログは、アプリケーションの全体的な正常性を理解するのに役立つだけでなく、より広範な問題をデバッグするときにも重要です。 アプリケーションは特定のエラーから復旧できますが、解決するためにはコード変更が必要な大規模な設計上の問題を示唆している可能性があります。 
> 
> このドキュメントで説明するエラー条件を実装する場合は、前述の理由でエラー コードと説明をログに記録する必要があります。 ログ記述コードの例については、「[エラーとログ記録のリファレンス](#error-and-logging-reference)」をご覧ください。 
>

## <a name="acquiretokensilent"></a>AcquireTokenSilent

AcquireTokenSilent は、エンド ユーザーにユーザー インターフェイス (UI) が表示されないことを保証してトークンを取得しようとします。 場合によっては、自動での取得が失敗することがあり、対話型の要求または既定のハンドラーによって処理する必要があります。 以降のセクションでは、各ケースを採用する場合と方法について詳しく説明します。

アプリケーションに固有のエラー処理が必要な可能性のある、オペレーティング システムによって生成される一連のエラーがあります。 詳しくは、「[エラーとログ記録のリファレンス](#error-and-logging-reference)」の「オペレーティング システム エラー」のセクションをご覧ください。 

### <a name="application-scenarios"></a>アプリケーションのシナリオ

- [ネイティブ クライアント](developer-glossary.md#native-client) アプリケーション (iOS、Android、.NET デスクトップ、または Xamarin)
- [リソース](developer-glossary.md#resource-server) (.NET) を呼び出す [Web クライアント](developer-glossary.md#web-client) アプリケーション

### <a name="error-cases-and-actionable-steps"></a>エラー ケースと実施可能な手順

基本的に、AcquireTokenSilent エラーには 2 つのケースがあります。

| ケース | 説明 |
|------|-------------|
| **ケース 1**: エラーは対話型のサインインで解決できる | 有効なトークンがないことが原因のエラーの場合は、対話型の要求が必要です。 具体的には、キャッシュ参照と無効/有効期限切れの更新トークンを解決するには AcquireToken 呼び出しが必要です。<br><br>このような場合は、エンド ユーザーにサインインを求める必要があります。 アプリケーションは対話型の要求をすぐに行うか、エンドユーザーの操作 ([サインイン] ボタンを押すなど) の後に行うか、またはそれ以降に行うかを選択できます。 選択は、アプリケーションの目的の動作によって決まります。<br><br>この特定のケースとそれを診断するエラーについては、次のセクションのコードをご覧ください。|
| **ケース 2**: エラーは対話型のサインインで解決できない | ネットワーク エラーと一時的なエラー、またはその他のエラーの場合は、対話型の AcquireToken 要求を実行しても問題は解決しません。 不必要な対話型サインインのプロンプトはエンド ユーザーにストレスを感じさせることもあります。 ADAL は、AcquireTokenSilent エラー発生時にほとんどのエラーについて再試行を自動的に 1 回行います。<br><br>クライアント アプリケーションは後で再試行してみることもできますが、実行するタイミングと方法は、アプリケーションの動作と必要なエンドユーザー エクスペリエンスによって決まります。 たとえば、アプリケーションは数分後に、またはなんらかのエンドユーザー アクションへの応答として AcquireTokenSilent の再試行を行うことができます。 すぐに再試行するとアプリケーションが制限されるため、試行しないでください。<br><br>後続の再試行が同じエラーで失敗しても、クライアントが AcquireToken を使って対話型の要求を行う必要があるということは意味しません。対話型の要求ではエラーは解決されません。<br><br>この特定のケースとそれを診断するエラーについては、次のセクションのコードをご覧ください。 |

### <a name="net"></a>.NET

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- acquireTokenSilentAsync(…)
- acquireTokenSilentSync(…) 
- [非推奨] acquireTokenSilent(…)
- [非推奨] acquireTokenByRefreshToken(…) 

コードは次のように実装されます。

```csharp
try{
    AcquireTokenSilentAsync(…);
} 

catch (AdalSilentTokenAcquisitionException e) {
    // Exception: AdalSilentTokenAcquisitionException
    // Caused when there are no tokens in the cache or a required refresh failed. 

    // Action: Case 1, resolvable with an interactive request. 
} 

catch(AdalServiceException e) {
    // Exception: AdalServiceException 
    // Represents an error produced by the STS. 
    // e.ErrorCode contains the error code and description, which can be used for debugging. 
    // NOTE: Do not code a dependency on the contents of the error description, as it can change over time.

    // Action: Case 2, not resolvable with an interactive request.
    // Attempt retry after a timed interval or user action.
} 
    
catch (AdalException e) {
    // Exception: AdalException 
    // Represents a library exception generated by ADAL .NET. 
    // e.ErrorCode contains the error code. 

    // Action: Case 2, not resolvable with an interactive request.
    // Attempt retry after a timed interval or user action.
    // Example Error: network_not_available, default case.
}
```

### <a name="android"></a>Android

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- acquireTokenSilentSync(…)
- acquireTokenSilentAsync(...)
- [非推奨] acquireTokenSilent(…)

コードは次のように実装されます。

```java
// *Inside callback*
public void onError(Exception e) {

    if (e instanceof AuthenticationException) {
        // Exception: AdalException
        // Represents a library exception generated by ADAL Android.
        // Error Code: e.getCode().

        // Errors: ADALError.ERROR_SILENT_REQUEST,
        // ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED,
        // ADALError.INVALID_TOKEN_CACHE_ITEM
        // Description: Request failed due to no tokens in
        // cache or failed a required refresh. 

        // Action: Case 1, resolvable with an interactive request. 

        // Action: Case 2, not resolvable with an interactive request.
        // Attempt retry after a timed interval or user action.
        // Example Errors: default case,
        // DEVICE_CONNECTION_IS_NOT_AVAILABLE, 
        // BROKER_AUTHENTICATOR_ERROR_GETAUTHTOKEN,
    }
}
```

### <a name="ios"></a>iOS

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- acquireTokenSilentWithResource(…)

コードは次のように実装されます。

```objc
[context acquireTokenSilentWithResource:[ARGS], completionBlock:^(ADAuthenticationResult *result) {
    if (result.status == AD_FAILED) {
        if ([error.domain isEqualToString:ADAuthenticationErrorDomain]){
            // Exception: AD_FAILED 
            // Represents a library error generated by ADAL Objective-C.
            // Error Code: result.error.code

            // Errors: AD_ERROR_SERVER_REFRESH_TOKEN_REJECTED, AD_ERROR_CACHE_NO_REFRESH_TOKEN
            // Description: No tokens in cache or failed a required token refresh failed. 
            // Action: Case 1, resolvable with an interactive request. 

            // Error: AD_ERROR_CACHE_MULTIPLE_USERS
            // Description: There was ambiguity in the silent request resulting in multiple cache items.
            // Action: Special Case, application should perform another silent request and specify the user using ADUserIdentifier. 
            // Can be caused in cases of a multi-user application. 

            // Action: Case 2, not resolvable with an interactive request.
            // Attempt retry after some time or user action.
            // Example Errors: default case,
            // AD_ERROR_CACHE_BAD_FORMAT
        }
    }
}]
```

## <a name="acquiretoken"></a>AcquireToken

AcquireToken は、トークンを取得するために使われる既定の ADAL メソッドです。 ユーザー ID が必要な場合、AcquireToken は最初にトークンを自動で取得しようとしてから、(PromptBehavior.Never が渡されていない限り) 必要に応じて UI を表示します。 アプリケーション ID が必要な場合、AcquireToken はトークンを取得しようとしますが、エンド ユーザーが存在しないため UI は表示しません。 

AcquireToken エラーを処理する場合、エラー処理は、プラットフォームと、アプリケーションが実現しようとしているシナリオに依存します。 

オペレーティング システムでは、特定のアプリケーションに依存するエラー処理を必要とする一連のエラーが生成されることもあります。 詳しくは、「[エラーとログ記録のリファレンス](#error-and-logging-reference)」の「オペレーティング システム エラー」をご覧ください。 

### <a name="application-scenarios"></a>アプリケーションのシナリオ

- ネイティブ クライアント アプリケーション (iOS、Android、.NET デスクトップ、または Xamarin)
- リソース API (.NET) を呼び出す Web アプリケーション
- シングル ページ アプリケーション (JavaScript)
- サービス間アプリケーション (.NET、Java)
  - On-Behalf-Of を含むすべてのシナリオ
  - On-Behalf-Of 固有のシナリオ

### <a name="error-cases-and-actionable-steps-native-client-applications"></a>エラー ケースと実施可能な手順: ネイティブ クライアント アプリケーション

ネイティブ クライアント アプリケーションを構築する場合は、ネットワークの問題、一時的なエラー、その他のプラットフォーム固有のエラーに関連して考慮すべきいくつかのエラー処理ケースがあります。 ほとんどの場合、アプリケーションではすぐに再試行を実行する必要はなく、サインインを求めるエンドユーザー操作を待機します。 

1 回の再試行で問題を解決できる特別なケースがいくつかあります。 たとえば、ユーザーがデバイス上のデータを有効にする必要がある場合や、最初のエラーの後に Azure AD ブローカーのダウンロードが完了した場合などです。 

エラーの場合、アプリケーションでは、エンド ユーザーが操作を実行できるように再試行を求める UI を表示できます。 たとえば、デバイスがオフライン エラーで失敗した場合は、失敗をすぐに再試行する代わりに AcquireToken の再試行を求める [もう一度サインインを試行] ボタンを表示できます。 

ネイティブ アプリケーションのエラー処理は 2 つのケースで定義できます。

|  |  |
|------|-------------|
| **ケース 1**:<br>再試行できないエラー (ほとんどの場合) | 1.すぐに再試行しないでください。 具体的なエラーに基づいて再試行を呼び出す UI ([もう一度サインインを試行]、[Download Azure AD broker application]\(Azure AD ブローカー アプリケーションをダウンロード\) など) をエンドユーザーに提示します。 |
| **ケース 2**:<br>再試行可能なエラー | 1.エンド ユーザーが成功する状態となった可能性があるため再試行を 1 回実行します。<br><br>2.再試行に失敗した場合は、具体的なエラーに基づいて再試行を呼び出す UI ([もう一度サインインを試行]、[Download Azure AD broker app]\(Azure AD ブローカー アプリをダウンロード\) など) をエンドユーザーに提示します。 |

> [!IMPORTANT]
> ユーザー アカウントがサイレント呼び出しで ADAL に渡され、失敗した場合、後続の対話型の要求でエンド ユーザーは別のアカウントを使ってサインインできます。 ユーザー アカウントを使って AcquireToken が成功した後、アプリケーションではサインイン ユーザーがアプリケーションのローカル ユーザー オブジェクトと一致することを確認する必要があります。 不一致があっても例外は生成されませんが (Objective C を除く)、認証要求の前にユーザーがローカルに既知であるケースを考慮する必要があります (失敗したサイレント呼び出しなど)。
>

#### <a name="net"></a>.NET

次のガイダンスは、以下を "*除く*" すべてのサイレントでない AcquireToken(…) ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- AcquireTokenAsync(…, IClientAssertionCertification, …)
- AcquireTokenAsync(…,ClientCredential, …)
- AcquireTokenAsync(…,ClientAssertion, …)
- AcquireTokenAsync(…,UserAssertion,…)   

コードは次のように実装されます。

```csharp
try {
    AcquireTokenAsync(…);
} 
    
catch(AdalServiceException e) {
    // Exception: AdalServiceException 
    // Represents an error produced by the STS. 
    // e.ErrorCode contains the error code and description, which can be used for debugging. 
    // NOTE: Do not code a dependency on the contents of the error description, as it can change over time.
    
    // Design time consideration: Certain errors may be caused at development and exposed through this exception. 
    // Looking inside the description will give more guidance on resolving the specific issue. 

    // Action: Case 1: Non-Retryable 
    // Do not perform an immediate retry. Only retry after user action. 
    // Example Errors: default case

    } 

catch (AdalException e) {
    // Exception: AdalException 
    // Represents a library exception generated by ADAL .NET.
    // e.ErrorCode contains the error code

    // Action: Case 1, Non-Retryable 
    // Do not perform an immediate retry. Only retry after user action. 
    // Example Errors: network_not_available, default case
}
```

> [!NOTE]
> ADAL .NET は、AcquireTokenSilent と同様に動作する PromptBehavior.Never をサポートしているため、追加の考慮事項があります。
>

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- acquireToken(…, PromptBehavior.Never)

コードは次のように実装されます。

```csharp
    try {acquireToken(…, PromptBehavior.Never);
    } 

catch(AdalServiceException e) {
    // Exception: AdalServiceException represents 
    // Represents an error produced by the STS. 
    // e.ErrorCode contains the error code and description, which can be used for debugging. 
    // NOTE: Do not code a dependency on the contents of the error description, as it can change over time.

    // Action: Case 1: Non-Retryable 
    // Do not perform an immediate retry. Only retry after user action. 
    // Example Errors: default case

} catch (AdalException e) {
    // Error Code: e.ErrorCode == "user_interaction_required"
    // Description: user_interaction_required indicates the silent request failed 
    // in a way that's resolvable with an interactive request.
    // Action: Resolvable with an interactive request. 

    // Action: Case 1, Non-Retryable 
    // Do not perform an immediate retry. Only retry after user action. 
    // Example Errors: network_not_available, default case
}
```

#### <a name="android"></a>Android

次のガイダンスは、すべてのサイレントでない AcquireToken(…) ADAL メソッドと組み合わせたエラー処理の例を示しています。 

コードは次のように実装されます。

```java
AcquireTokenAsync(…);

// *Inside callback*
public void onError(Exception e) {
    if (e instanceof AuthenticationException) {
        // Exception: AdalException 
        // Represents a library exception generated by ADAL Android.
        // Error Code: e.getCode();

        // Error: ADALError.BROKER_APP_INSTALLATION_STARTED
        // Description: Broker app not installed, user will be prompted to download the app. 

        // Action: Case 2, Retriable Error 
        // Perform a single retry. If that fails, only try again after user action. 

        // Action: Case 1, Non-Retriable 
        // Do not perform an immediate retry. Only retry after user action. 
        // Example Errors: default case, DEVICE_CONNECTION_IS_NOT_AVAILABLE
    }
}
```

#### <a name="ios"></a>iOS

次のガイダンスは、すべてのサイレントでない AcquireToken(…) ADAL メソッドと組み合わせたエラー処理の例を示しています。 

コードは次のように実装されます。

```objc
[context acquireTokenWithResource:[ARGS], completionBlock:^(ADAuthenticationResult *result) {
    if (result.status == AD_FAILED) {
        if ([error.domain isEqualToString:ADAuthenticationErrorDomain]){
            // Exception: AD_FAILED 
            // Represents a library error generated by ADAL ObjC.
            // Error Code: result.error.code 

            // Error: AD_ERROR_SERVER_WRONG_USER
            // Description: App passed a user into ADAL and the end user signed in with a different account. 
            // Action: Case 1, Non-retriable (as is) and up to the application on how to handle this case. 
            // It can attempt a new request without specifying the user, or use UI to clarify the user account to sign in. 

            // Action: Case 1, Non-Retriable 
            // Do not perform an immediate retry. Only retry after user action. 
            // Example Errors: default case
        }
    }
}]
```

### <a name="error-cases-and-actionable-steps-web-applications-that-call-a-resource-api-net"></a>エラー ケースと実施可能な手順: リソース API (.NET) を呼び出す Web アプリケーション

リソースの承認コードを使ったトークンの取得を呼び出す .NET Web アプリを構築する場合、必要なコードは汎用ケースの既定のハンドラーだけです。 

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- AcquireTokenByAuthorizationCodeAsync(…)

コードは次のように実装されます。

```csharp
try {
    AcquireTokenByAuthorizationCodeAsync(…);
} 

catch (AdalException e) {
    // Exception: AdalException
    // Represents a library exception generated by ADAL .NET.
    // Error Code: e.ErrorCode

    // Action: Do not perform an immediate retry. Only try again after user action or wait until much later. 
    // Example Errors: default case
}
```

### <a name="error-cases-and-actionable-steps-single-page-applications-adaljs"></a>エラー ケースと実施可能な手順: 単一ページ アプリケーション (adal.js)

AcquireToken で adal.js を使って単一ページ アプリケーションを構築する場合、エラー処理コードは通常のサイレント呼び出しのコードと同様です。 具体的には、adal.js では AcquireToken は UI を決して表示しません。 

失敗した AcquireToken には次のケースがあります。

|  |  |
|------|-------------|
| **ケース 1**:<br>対話型の要求で解決可能 | 1.login() に失敗した場合は、すぐに再試行を実行しないでください。 ユーザー操作で再試行が要求された後でのみ再試行します。|
| **ケース 2**:<br>対話型の要求で解決できない。 エラーは再試行可能。 | 1.エンド ユーザーが成功する状態となった可能性があるため再試行を 1 回実行します。<br><br>2.再試行に失敗した場合は、具体的なエラーに基づいて再試行を呼び出せるアクション ([もう一度サインインを試行]) をエンド ユーザーに提示します。 |
| **ケース 3**:<br>対話型の要求で解決できない。 エラーは再試行可能ではない。 | 1.すぐに再試行しないでください。 具体的なエラーに基づいて再試行を呼び出せるアクション ([もう一度サインインを試行]) をエンド ユーザーに提示します。 |

コードは次のように実装されます。

```javascript
AuthContext.acquireToken(…, function(error, errorDesc, token) {
    if (error || errorDesc) {
        // Represents any token acquisition failure that occurred. 
        // Error Code: error.indexOf("<ERROR_STRING>")

        // Errors: if (error.indexOf("interaction_required"))
        //         if (error.indexOf("login required"))
        // Description: ADAL wasn't able to silently acquire a token because of expire or fresh session. 
        // Action: Case 1, Resolvable with an interactive login() request. 

        // Error: if (error.indexOf("Token Renewal Failed")) 
        // Description: Timeout when refreshing the token.
        // Action: Case 2, Not resolvable interactively, error is retriable.
        // Perform a single retry. Only try again after user action.

        // Action: Case 3, Not resolvable interactively, error is not retriable. 
        // Do not perform an immediate retry. Only retry after user action.
        // Example Errors: default case
    }
}
```

### <a name="error-cases-and-actionable-steps-service-to-service-applications-net-only"></a>エラー ケースと実施可能な手順: サービス間アプリケーション (.NET のみ)

AcquireToken を使うサービス間アプリケーションを構築する場合は、コードで処理する必要がある重要なエラーがいくつかあります。 エラーに対する手段は、呼び出し元のアプリにエラーを返すか (On-Behalf-Of の場合)、再試行戦略を適用することだけです。 

#### <a name="all-scenarios"></a>すべてのシナリオ

On-Behalf-Of を含む "*すべて*" のサービス間アプリケーション シナリオの場合:

- すぐに再試行しないでください。 ADAL は特定の失敗した要求に対して再試行を 1 回行います。 
- ユーザーまたはアプリのアクションで再試行が要求された後でのみ再試行に進みます。 たとえば、設定された間隔で動作するデーモン アプリケーションは、次の間隔まで待ってから再試行する必要があります。

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- AcquireTokenAsync(…, IClientAssertionCertification, …)
- AcquireTokenAsync(…,ClientCredential, …)
- AcquireTokenAsync(…,ClientAssertion, …)
- AcquireTokenAsync(…,UserAssertion, …)

コードは次のように実装されます。

```csharp
try {
    AcquireTokenAsync(…);
} 

catch (AdalException e) {
    // Exception: AdalException
    // Represents a library exception generated by ADAL .NET.
    // Error Code: e.ErrorCode

    // Action: Do not perform an immediate retry. Only try again after user action (if applicable) or wait until much later. 
    // Example Errors: default case
}  
```

#### <a name="on-behalf-of-scenarios"></a>On-behalf-Of のシナリオ

*On-behalf-Of* サービス間アプリケーション シナリオの場合。

次のガイダンスは、ADAL メソッドと組み合わせたエラー処理の例を示しています。 

- AcquireTokenAsync(…, UserAssertion, …)

コードは次のように実装されます。

```csharp
try {
AcquireTokenAsync(…);
} 

catch (AdalServiceException e) {
    // Exception: AdalServiceException 
    // Represents an error produced by the STS. 
    // e.ErrorCode contains the error code and description, which can be used for debugging. 
    // NOTE: Do not code a dependency on the contents of the error description, as it can change over time.

    // Error: On-Behalf-Of Error Handler
    if (e.ErrorCode == "interaction_required") {
        // Description: The client needs to perform some action due to a config from admin. 
        // Action: Capture `claims` parameter inside ex.InnerException.InnerException.
        // Generate HTTP error 403 with claims, throw back HTTP error to client.
        // Wait for client to retry. 
    }
} 
        
catch (AdalException e) {
    // Exception: AdalException 
    // Represents a library exception generated by ADAL .NET.
    // Error Code: e.ErrorCode

    // Action: Do not perform an immediate retry. Only try again after user action (if applicable) or wait until much later. 
    // Example Error: default case
}
```

このシナリオを示す[完全なサンプル](https://github.com/Azure-Samples/active-directory-dotnet-webapi-onbehalfof-ca)を作成しました。

## <a name="error-and-logging-reference"></a>エラーとログ記録のリファレンス

### <a name="logging-personal-identifiable-information-pii--organizational-identifiable-information-oii"></a>組織を特定できる情報 (OII) と個人を特定できる情報 (PII) のログ記録
既定では、ADAL ログには PII や OII は取得または記録されません。 アプリ開発者はライブラリの Logger クラスの setter から、この設定をオンにできます。 PII または OII の設定をオンにすると、アプリは、規制の要件に準拠して、高度な機密データを安全に処理する責任を負います。

### <a name="net"></a>.NET

#### <a name="adal-library-errors"></a>ADAL ライブラリ エラー

特定の ADAL エラーについて調べるには、[azure-activedirectory-library-for-dotnet リポジトリ](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/8f6d560fbede2247ec0e217a21f6929d4375dcaa/src/ADAL.PCL/Utilities/Constants.cs#L58)のソース コードが最適なエラー リファレンスです。

#### <a name="guidance-for-error-logging-code"></a>エラーのログ記録コードに関するガイダンス

ADAL .NET のログ記録は、作業しているプラットフォームに応じて変わります。 ログ記録を有効にするコードについては、[ログ記録 Wiki](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Logging-in-ADAL.Net) をご覧ください。

### <a name="android"></a>Android

#### <a name="adal-library-errors"></a>ADAL ライブラリ エラー

特定の ADAL エラーについて調べるには、[azure-activedirectory-library-for-android リポジトリ](https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/dev/adal/src/main/java/com/microsoft/aad/adal/ADALError.java#L33)のソース コードが最適なエラー リファレンスです。

#### <a name="operating-system-errors"></a>オペレーティング システム エラー

ADAL の AuthenticationException を通じて公開されている Android OS エラーは "SERVER_INVALID_REQUEST" として識別可能で、エラーの説明を通じてさらに細かく分類できます。 

よくあるエラーと、アプリやエンドユーザーがそれに遭遇した場合の対策の詳細一覧については、[ADAL Android Wiki ](https://github.com/AzureAD/azure-activedirectory-library-for-android/wiki)を参照してください。 

#### <a name="guidance-for-error-logging-code"></a>エラーのログ記録コードに関するガイダンス

```java
// 1. Configure Logger
Logger.getInstance().setExternalLogger(new ILogger() {    
    @Override   
    public void Log(String tag, String message, String additionalMessage, LogLevel level, ADALError errorCode) { 
    // …
    // You can write this to logfile depending on level or errorcode. 
    writeToLogFile(getApplicationContext(), tag +":" + message + "-" + additionalMessage);    
    }
}

// 2. Set the log level
Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);

// By default, the `Logger` does not capture any PII or OII

// PII or OII will be logged
Logger.getInstance().setEnablePII(true);

// To STOP logging PII or OII, use the following setter
Logger.getInstance().setEnablePII(false);


// 3. Send logs to logcat.
adb logcat > "C:\logmsg\logfile.txt";
```

### <a name="ios"></a>iOS

#### <a name="adal-library-errors"></a>ADAL ライブラリ エラー

特定の ADAL エラーについて調べるには、[azure-activedirectory-library-for-objc リポジトリ](https://github.com/AzureAD/azure-activedirectory-library-for-objc/blob/dev/ADAL/src/ADAuthenticationError.m#L295)のソース コードが最適なエラー リファレンスです。

#### <a name="operating-system-errors"></a>オペレーティング システム エラー

iOS エラーは、サインイン時にユーザーが Web ビューを使っているとき、および認証の性質として発生することがあります。 これは、SSL エラー、タイムアウト、ネットワーク エラーなどの条件によって発生することがあります。

- 権利の共有では、ログインは永続的ではなく、キャッシュが空のように見えます。 キー チェーンに次のコード行を追加すると解決できます。`[[ADAuthenticationSettings sharedInstance] setSharedCacheKeychainGroup:nil];`
- NsUrlDomain の一連のエラーでは、アクションはアプリのロジックに応じて変わります。 処理できる具体的な例については、[NSURLErrorDomain のリファレンス ドキュメント](https://developer.apple.com/documentation/foundation/nsurlerrordomain#declarations)をご覧ください。
- ADAL Objective-C チームが保守している一般的なエラーの一覧については、[ADAL Obj-C の一般的な問題](https://github.com/AzureAD/azure-activedirectory-library-for-objc#adauthenticationerror)に関するページをご覧ください。

#### <a name="guidance-for-error-logging-code"></a>エラーのログ記録コードに関するガイダンス

```objc
// 1. Enable NSLogging
[ADLogger setNSLogging:YES];

// 2. Set the log level (if you want verbose)
[ADLogger setLevel:ADAL_LOG_LEVEL_VERBOSE];

// 3. Set up a callback block to simply print out
[ADLogger setLogCallBack:^(ADAL_LOG_LEVEL logLevel, NSString *message, NSString *additionalInformation, NSInteger errorCode, NSDictionary *userInfo) {
     NSString* log = [NSString stringWithFormat:@"%@ %@", message, additionalInformation];    
     NSLog(@"%@", log);
}];
```

### <a name="guidance-for-error-logging-code---javascript"></a>エラーのログ記録コードに関するガイダンス - JavaScript 

```javascript
0: Error1: Warning2: Info3: Verbose
window.Logging = {
    level: 3,
    log: function (message) {
        console.log(message);
    }
};
```
## <a name="related-content"></a>関連コンテンツ

* [Azure AD 開発者ガイド][AAD-Dev-Guide]
* [Azure AD Authentication ライブラリ][AAD-Auth-Libraries]
* [Azure AD Authentication のシナリオ][AAD-Auth-Scenarios]
* [Azure Active Directory とアプリケーションの統合][AAD-Integrating-Apps]

Microsoft のコンテンツ改善のため、以下のコメント セクションよりご意見をお寄せください。

[![サインイン ボタン][AAD-Sign-In]][AAD-Sign-In]
<!--Reference style links --> [AAD-Auth-Libraries]: ./active-directory-authentication-libraries.md [AAD-Auth-Scenarios]:authentication-scenarios.md [AAD-Dev-Guide]:azure-ad-developers-guide.md [AAD-Integrating-Apps]:quickstart-v1-integrate-apps-with-azure-ad.md [AZURE-portal]: https://portal.azure.com

<!--Image references-->
[AAD-Sign-In]:./media/active-directory-devhowto-multi-tenant-overview/sign-in-with-microsoft-light.png

